Add context options: -A/-B/-C; Fixes #287

flopp opened the pull request:
This PR adds the options `-A/--after-context`, `-B/--before-context`, and `-C/--context` that enable the display of context lines surrounding a detected spelling error. Using `-C LINES` results in the display of LINES many context lines before and after the erroneous line, with `-B` and `-A` the before and after context can be configured separately. The default is to not show any context at all (so there's no change in behaviour if no context options are used).

luzpaz commented:
ooh I can't wait to test this! Thanks!

larsoner commented:
> why is flake8 suddenly complaining about the (invalid) word_regex_def that's not touched by this commit?

As the upstream libraries get updated, they can become more strict.

> `word_regex_def = u"[\\w\\-'’`]+"`

I would go with this one because it might need to be `unicode` in Python2.7 to be concatenated. Though I agree the `r""` version looks better so feel free to try it and we can see if the CIs are happy.

Can you add a few simple tests, including that things like `-A 1 -B1` with a one-line file does not break everything?

flopp commented:
Ok, I've added some tests for various combinations of the context options (including the case that more context is requested than available in the file). The tests actually revealed an error when only a "before context" is requested - I fixed that.

larsoner commented:
Horray unit testing!

@luzpaz can you give this PR a try?

luzpaz commented:
> @luzpaz can you give this PR a try?

So, I'm triyng to test it.. is it not enough to checkout the PR and run codespell ? Do i need to install it ? 
I'm running `codespell -h` and don't see the new flags 
```
$ codespell -A
Usage: 
	codespell [OPTIONS] [file1 file2 ... fileN]


codespell: error: no such option: -A
```

larsoner commented:
It depends on how you installed. If you did `python setup.py develop` or `pip install -e .` then yes checking out should be enough.

luzpaz commented:
@larsoner I did neither :blush: 
I just checked out the branch. 
I install codespell via pip generally 
`pip install --user --upgrade git+https://github.com/codespell-project/codespell.git`
So how do you recommend I go about testing this PR ?

larsoner commented:
Checkout the repo to your local machine, install in editable mode, check out this PR with some commands like (only some of which you probably need):
```
$ git checkout git@github.com:/codespell-project/codespell
$ cd codespell
$ pip install --user -e .
$ git fetch origin pull/722/head:722
$ which codespell
$ codespell -A 2 codespell_lib/tests/
```
Then whatever you check out at the time will be what `codespell` runs.

luzpaz commented:
It looks like it works!
Nice!
Green light from me. 

@larsoner Do i need to uninstall anything after I'm done testing ? Or just remove the 722 branch ?

luzpaz commented:
Actually @flopp it looks like you've diverted from `grep` behavior. 
As specified in #287 in `grep`:
> `grep -A <number>` returns the `<number>` of lines **before** the match
> `grep -B <number>` returns the `<number>` of lines **after** the match
> `grep -C <number>` returns the `<number>` of lines **before AND after** the match

In this PR 
> `-A/--after-context`, `-B/--before-context`, and `-C/--context`

So `-A` should really be **before** 
and `-B` should be **after**

larsoner commented:
> Do i need to uninstall anything after I'm done testing ? Or just remove the 722 branch ?

No you can just do:
```
git checkout master
```
and to update later do
```
git pull origin master
```

flopp commented:
@luzpaz I'm confident that #287 is wrong - and the implementation in this PR is consistent with grep. Grep's manual (https://www.gnu.org/software/grep/manual/grep.html#Context-Line-Control) clearly  states that '-A' is the 'after context' and '-B' is the 'before context'.

luzpaz commented:
@flopp oops.. that was prob. a moment of dyslexia. Cancel clear. You are correct. Green light from me

larsoner commented on codespell_lib/_codespell.py:
```diff
@@ -411,8 +417,20 @@ def ask_for_word_fix(line, wrongword, misspelling, interactivity):
+            if i == index:
+                print('> %s' % lines[i].rstrip())
+            else:
+                print(': %s' % lines[i].rstrip())
```
Perhaps could be slightly more Pythonic?
```
    for i in range(max(index - context_before, 0),
                   min(index + context_after + 1, len(lines))):
        print('%s %s' % ('>' if i == index else ':', lines[i].rstrip())
```

flopp replied:
You're right. Let me refactor the code.

larsoner commented on codespell_lib/_codespell.py:
```diff
@@ -411,8 +417,20 @@ def ask_for_word_fix(line, wrongword, misspelling, interactivity):
 
+def print_context(lines, index, context_before, context_after):
+    if context_before < 0:
+        return
```
How can `context_before` ever be `< 0`? This seems like it should not happen. If so, inputs should be sanity checked / fixed so this conditional is no longer necessary.

flopp replied:
`context_before` is negative if no context option is used at all. But I see that this may be confusing. I'll try to refactor the code a bit.

larsoner commented on codespell_lib/tests/test_basic.py:
```diff
@@ -302,6 +302,68 @@ def test_case_handling(tmpdir, capsys):
+        lines = capsys.readouterr()[0].split('\n')
+        assert 'ERROR' in lines[0]
+    finally:
+        os.remove(f.name)
```
The beauty of `tmpdir` fixture is that you do not need to do this `try/except` business, `pytest` should take care of removal for you

flopp replied:
Ok, I see. I just followed the other test functions, that all include the `try/finally` pattern. So, I don't need to call `os.remove(f.name)` explicitly?

peternewman commented:
> > why is flake8 suddenly complaining about the (invalid) word_regex_def that's not touched by this commit?
> 
> As the upstream libraries get updated, they can become more strict.
> 
> > `word_regex_def = u"[\\w\\-'’`]+"`
> 
> I would go with this one because it might need to be `unicode` in Python2.7 to be concatenated. Though I agree the `r""` version looks better so feel free to try it and we can see if the CIs are happy.
> 

I hit this on another project recently (due to the same Python update). This says you can use ur for the best of both worlds:
https://docs.python.org/2/reference/lexical_analysis.html#string-literals

`word_regex_def = ur"[\w\-'’`]+"`

larsoner commented:
My sense is that we should merge this and then tackle #726, WDYT @flopp ?

flopp commented:
@larsoner there seems to be a problem with `word_regex_def = ur"..."` and Python 3.X (see travis output for #735).

Other than that I'm ok with merging the current version of the PR.

larsoner commented:
Thanks @flopp

luzpaz commented:
Thank you @flopp and everyone else!
